builtin/logical/pki/cert_util.go

```diff
@@ -1513,3 +1513,288 @@ func stringToOid(in string) (asn1.ObjectIdentifier, error) {
+
+		parentSubject := string(issuerIdCertMap[child].RawIssuer)
+		parentCerts, ok := subjectIssuerIdsMap[parentSubject]
+		if !ok {
+			// When the issuer isn't known to Vault, the lookup by the issuer
+			// will be empty. This most commonly occurs when intermediates are
+			// directly added (via intermediate/set-signed) without providing
+			// the root.
+			continue
+		}
+
+		// Now, iterate over all possible parents and assign the child/parent
+		// relationship.
+		for _, parent := range parentCerts {
+			// Skip self-references to the exact same certificate.
+			if child == parent {
+				continue
+			}
+
+			// While we could use Subject/Authority Key Information (SPKI/APKI)
+			// as a heuristic for whether or not this relationship is valid,
+			// this is insufficient as otherwise valid CA certificates could
+			// elide this information. That means its best to actually validate
+			// the signature (e.g., call child.CheckSignatureFrom(parent))
+			// instead.
+			parentCert := issuerIdCertMap[parent]
+			if err := childCert.CheckSignatureFrom(parentCert); err != nil {
+				return nil
+			}
+
+			// Otherwise, we can append it to the map, allowing us to walk the
+			// issuer->parent mapping.
+			issuerIdParentsMap[child] = append(issuerIdParentsMap[child], parent)
+
+			// Also cross-add the child relationship step at the same time.
+			issuerIdChildrenMap[parent] = append(issuerIdChildrenMap[parent], child)
+		}
+	}
+
+	// Finally, we consult RFC 8446 Section 4.4.2 for creating an algorithm for
+	// building the chain:
+	//
+	// > ... The sender's certificate MUST come in the first
+	// > CertificateEntry in the list.  Each following certificate SHOULD
+	// > directly certify the one immediately preceding it.  Because
+	// > certificate validation requires that trust anchors be distributed
+	// > independently, a certificate that specifies a trust anchor MAY be
+	// > omitted from the chain, provided that supported peers are known to
+	// > possess any omitted certificates.
+	// >
+	// > Note: Prior to TLS 1.3, "certificate_list" ordering required each
+	// > certificate to certify the one immediately preceding it; however,
+	// > some implementations allowed some flexibility.  Servers sometimes
+	// > send both a current and deprecated intermediate for transitional
+	// > purposes, and others are simply configured incorrectly, but these
+	// > cases can nonetheless be validated properly.  For maximum
+	// > compatibility, all implementations SHOULD be prepared to handle
+	// > potentially extraneous certificates and arbitrary orderings from any
+	// > TLS version, with the exception of the end-entity certificate which
+	// > MUST be first.
+	//
+	// So, we take this to mean we should build chains via DFS: each issuer is
+	// explored until an empty parent pointer (i.e., self-loop) is reached and
+	// then latest-most duplicate parent is then explored.
+	//
+	// However, we don't actually need to do a DFS (per issuer) here. We can
```

Comment:
I think just means it is a topographical sort, making it `O(|V| + |E|)`, which means the entire algorithm is `O(n^2)` in the number of certificates (either by the toposort or by the above construction).